radixdlt-core/radixdlt/src/integration/java/com/radixdlt/integration/api/ApiTest.java

```diff
@@ -104,38 +110,96 @@
+import com.radixdlt.ledger.LedgerUpdate;
 import com.radixdlt.mempool.MempoolConfig;
+import com.radixdlt.mempool.MempoolRelayTrigger;
 import com.radixdlt.network.p2p.PeersView;
+import com.radixdlt.statecomputer.InvalidProposedTxn;
 import com.radixdlt.statecomputer.checkpoint.MockedGenesisModule;
+import com.radixdlt.application.system.FeeTable;
+import com.radixdlt.statecomputer.forks.ForkOverwritesWithShorterEpochsModule;
 import com.radixdlt.statecomputer.forks.ForksModule;
 import com.radixdlt.statecomputer.forks.RERulesConfig;
 import com.radixdlt.statecomputer.forks.RadixEngineForksLatestOnlyModule;
 import com.radixdlt.store.DatabaseEnvironment;
 import com.radixdlt.store.DatabaseLocation;
 import com.radixdlt.store.berkeley.BerkeleyLedgerEntryStore;
-import com.radixdlt.utils.UInt256;
+import com.radixdlt.sync.messages.local.SyncCheckTrigger;
 
+import java.util.Collection;
 import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Random;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
-import static com.radixdlt.api.data.TransactionStatus.TRANSACTION_NOT_FOUND;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.Assert.assertEquals;
+/**
+ * Test which runs a 20 node consensus network. Random transactions are submitted and nodes
+ * are rebooted while checks occur on the api to make sure that invariants are not broken.
+ */
+@RunWith(Parameterized.class)
+public class ApiTest {
+	private static final Logger logger = LogManager.getLogger();
+	private static final int ACTION_ROUNDS = 5000;
+	private static final RERulesConfig config = RERulesConfig.testingDefault().overrideMaxSigsPerRound(2);
+	private static final Amount PER_BYTE_FEE = Amount.ofMicroTokens(2);
+	private static final List<ActorConfiguration> ACTOR_CONFIGURATIONS = List.of(
+		new ActorConfiguration(ApiTxnSubmitter::new, 1, 2),
+		new ActorConfiguration(BalanceReconciler::new, 1, 10),
+		new ActorConfiguration(RandomNodeRestarter::new, 1, 10),
+		new ActorConfiguration(NativeTokenRewardsChecker::new, 1, 100),
+		new ActorConfiguration(ApiBalanceToRadixEngineChecker::new, 1, 200)
+	);
+
+	@Parameterized.Parameters
+	public static Collection<Object[]> forksModule() {
+		return List.of(new Object[][]{
+			{new RadixEngineForksLatestOnlyModule(config.overrideMaxRounds(100)), null},
+			{new ForkOverwritesWithShorterEpochsModule(config), null},
+			{
+				new ForkOverwritesWithShorterEpochsModule(config),
+				new ForkOverwritesWithShorterEpochsModule(config.removeSigsPerRoundLimit())
+			},
+			{
+				new RadixEngineForksLatestOnlyModule(
+					config.overrideMaxRounds(100).overrideFeeTable(
+						FeeTable.create(
+							PER_BYTE_FEE,
+							Map.of()
+						)
+					)
+				), null},
+			{
+				new ForkOverwritesWithShorterEpochsModule(
+					config.overrideFeeTable(
+						FeeTable.create(
+							PER_BYTE_FEE,
+							Map.of()
+						)
+					)
+				), null},
+		});
```

Comment:
Do you think there is a way to make this block a bit more readable?

Reply:
Hard to do so. I think right way to fix this is to use junit5 parametrized tests. But that's for another PR.